File: Solved ques/maxpathsum.cpp
#include <iostream>
#include <climits>
#include <algorithm>
using namespace std;

// Binary Tree Node Structure
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;

    TreeNode(int x) {
        val = x;
        left = right = NULL;
    }
};

class Solution {
public:
    // Recursive function to compute max path sum
    int maxPath(TreeNode* node, int& maxi) {
        if (node == NULL) return 0;

        // Recursively find max sum from left and right
        int left = max(0, maxPath(node->left, maxi));  // ignore negative paths
        int right = max(0, maxPath(node->right, maxi));

        // Current node path sum (left + right + node)
        int currPath = left + right + node->val;

        // Update global maximum if current is greater
        maxi = max(maxi, currPath);

        // Return the maximum one-side path for parent
        return node->val + max(left, right);
    }

    // Main function to call
    int maxPathSum(TreeNode* root) {
        int maxi = INT_MIN;
        maxPath(root, maxi);
        return maxi;
    }
};

int main() {
    // Create tree:
    //       -10
    //       /  \
    //      9    20
    //          /  \
    //         15   7

    TreeNode* root = new TreeNode(-10);
    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);

    Solution sol;
    int result = sol.maxPathSum(root);
    cout << "Maximum Path Sum: " << result << endl;

    return 0;
}

File: Trees/MaxPathSum.java
package Trees;

public class MaxPathSum {
    
     public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {}

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.right = right;
            this.left = left;
        }
    }
    public static void main(String[] args) {
    MaxPathSum obj = new MaxPathSum();
    
    // Build a tree:
    //        -10
    //       /   \
    //      9     20
    //           /  \
    //          15   7
    //
    TreeNode root = obj.new TreeNode(-10);
    root.left = obj.new TreeNode(9);
    root.right = obj.new TreeNode(20);
    root.right.left = obj.new TreeNode(15);
    root.right.right = obj.new TreeNode(7);
    
    int result = obj.MaxPathSum(root);
    System.out.println("Maximum Path Sum: " + result);
}

    public int MaxPathSum(TreeNode root){
        int maxValue[]=new int[1];
        maxValue[0]=Integer.MIN_VALUE;
        MaxPathDown(root, maxValue);
        return maxValue[0];
    }

    public int MaxPathDown(TreeNode root,int maxValue[]){
        if(root==null) return 0;
        int left=Math.max(0,MaxPathDown(root.left, maxValue));
        int right=Math.max(0,MaxPathDown(root.right, maxValue));
        maxValue[0]=Math.max(maxValue[0], left+right+root.val);
        return Math.max(left,right)+root.val;
    }

}
